### PR TITLE
Preemptible gcp vms

### PR DESCRIPTION
**Pull request by noahl**

### CODE REVIEW

File: tests/gce_virtual_machine_test.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+from perfkitbenchmarker import flags
+from perfkitbenchmarker import virtual_machine
+from perfkitbenchmarker import vm_util
+from perfkitbenchmarker.gcp import gce_virtual_machine
+
+
+class GCEPreemptibleVMFlagTestCase(unittest.TestCase):
+  def testPreemptibleVMFlag(self):
+    with mock.patch(vm_util.__name__ + '.IssueCommand') as issueCommand, \
+            mock.patch('__builtin__.open'), \
```
Comment (cmccoy) [inline tests/gce_virtual_machine_test.py]: Why do you need to patch `open` and `NamedTemporaryFile` here?

File: tests/gce_virtual_machine_test.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            mock.patch(vm_util.__name__ + '.NamedTemporaryFile'):
+      try:
+        old_preemptible_flag = flags.FLAGS.gce_preemptible_vms
```
Comment (cmccoy) [inline tests/gce_virtual_machine_test.py]: Prefer patching `gce_virtual_machine.FLAGS`.

File: tests/gce_virtual_machine_test.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+from perfkitbenchmarker import pkb  # noqa
+from perfkitbenchmarker import flags
+from perfkitbenchmarker import virtual_machine
+from perfkitbenchmarker import vm_util
+from perfkitbenchmarker.gcp import gce_virtual_machine
+
+
+class GCEPreemptibleVMFlagTestCase(unittest.TestCase):
+  def testPreemptibleVMFlag(self):
+    with mock.patch(vm_util.__name__ + '.IssueCommand') as issueCommand, \
```
Comment (cmccoy) [inline tests/gce_virtual_machine_test.py]: Use `_`-delimited local variable names.

**Comment by voellm** [comment]
Noah - Whats the status of this change?  Have you addressed Connors feedback?

File: tests/gce_virtual_machine_test.py (hunk 3 above)
Comment (noahl) [inline tests/gce_virtual_machine_test.py]: Done.

File: tests/gce_virtual_machine_test.py (hunk 1 above)
Comment (noahl) [inline tests/gce_virtual_machine_test.py]: Because GceVirtualMachine._Create() uses those methods.

Comment (cmccoy) [inline tests/gce_virtual_machine_test.py]: Will creating a temporary file in your test cause issues? If not, might be more realistic to leave them.

File: tests/gce_virtual_machine_test.py (hunk 2 above)
Comment (noahl) [inline tests/gce_virtual_machine_test.py]: Done. I just pushed an update which I think addresses all of your comments.

File: tests/gce_virtual_machine_test.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for perfkitbenchmarker.gcp.gce_virtual_machine"""
+
+import unittest
+import mock
+
+from perfkitbenchmarker import pkb  # noqa
```
Comment (cmccoy) [inline tests/gce_virtual_machine_test.py]: Nit: Add a comment indicating why this module was imported.

**Comment by cmccoy** [comment]
Is there any indication on the resulting samples that preemptible VMs were used? Otherwise, LGTM.

File: tests/gce_virtual_machine_test.py (hunk 4 above)
Comment (noahl) [inline tests/gce_virtual_machine_test.py]: Done.

**Comment by noahl** [comment]
I don't think there will be any indication in the samples that the VMs were preemptible. I see two ways to accomplish that. The first option is to record data under fake machine types, like n1-standard-8-preemptible or something like that. I think I could make that change locally to gce_virtual_machine.py. The other option is to make it a flag in the metadata. In the second case, I think I would need to add a new interface for a VM to provide metadata to the publisher. It wouldn't be a huge change, but it would be bigger than the first option.

I don't have strong feelings about this. I think it depends on whether we think preemptible images are sufficiently different than regular ones that they should be considered their own type of machine, and I don't know the answer to that. What do you think?

**Comment by noahl** [comment]
I just uploaded a new commit that annotates samples from GCP VMs with a metadata label 'preemptible'. Possible values are True and False.

File: perfkitbenchmarker/gcp/gce_virtual_machine.py [hunk 5]
```diff
@@ -82,6 +85,13 @@ def __init__(self, vm_spec):
     self.max_local_disks = FLAGS.gce_num_local_ssds
     self.boot_metadata = {}
 
+    if FLAGS.gce_preemptible_vms:
```
Comment (cmccoy) [inline perfkitbenchmarker/gcp/gce_virtual_machine.py]: self.preemptible = FLAGS.gce_preemptible_vms?

File: perfkitbenchmarker/gcp/gce_virtual_machine.py [hunk 6]
```diff
@@ -226,6 +238,12 @@ def AddMetadata(self, **kwargs):
     cmd.extend(util.GetDefaultGcloudFlags(self))
     vm_util.IssueCommand(cmd)
 
+  def AnnotateSample(self, unused_sender, benchmark_spec, sample):
+    if self.preemptible:
```
Comment (cmccoy) [inline perfkitbenchmarker/gcp/gce_virtual_machine.py]: No need to switch here.
`sample['metadata']['preemptible'] = self.preemptible`

**Comment by cmccoy** [comment]
Two nits, otherwise LGTM.

File: perfkitbenchmarker/gcp/gce_virtual_machine.py (hunk 5 above)
Comment (noahl) [inline perfkitbenchmarker/gcp/gce_virtual_machine.py]: Done.

File: perfkitbenchmarker/gcp/gce_virtual_machine.py (hunk 6 above)
Comment (noahl) [inline perfkitbenchmarker/gcp/gce_virtual_machine.py]: Done.

**Pull request merged**